src/assets/scripts/client/canvas/CanvasController.js

```diff
@@ -132,108 +321,127 @@ export default class CanvasController {
         this.$window = null;
         this.$element = null;
         this.canvas = {};
-        this.canvas.contexts = {};
+        this._context = {};
         this.canvas.panY = 0;
         this.canvas.panX = 0;
         // resize canvas to fit window?
-        this.canvas.resize = true;
+        this._shouldResize = true;
         // all canvases are the same size
         this.canvas.size = {
             height: DEFAULT_CANVAS_SIZE.HEIGHT,
             width: DEFAULT_CANVAS_SIZE.WIDTH
         };
-        this.canvas.last = TimeKeeper.gameTime;
-        this.canvas.dirty = true;
-        this.canvas.draw_labels = false;
-        this.canvas.draw_restricted = false;
-        this.canvas.draw_sids = false;
-        this.canvas.draw_terrain = true;
+        this._lastFrameTimestamp = TimeKeeper.gameTime;
+        this._shouldShallowRender = true;
+        this._shouldDeepRender = true;
+        this._shouldDrawFixLabels = false;
+        this._shouldDrawRestrictedAreas = false;
+        this._shouldDrawSidMap = false;
+        this._shouldDrawTerrain = true;
 
         return this;
     }
 
     /**
-     * @for CanvasController
-     * @method canvas_init_pre
-     */
-    canvas_init_pre() {
-        return this;
-    }
-
-    /**
+     * Called by `AppController.init()`
+     *
+     * Creates canvas elements and stores context
+     *
      * @for CanvasController
      * @method canvas_init
      */
     canvas_init() {
-        this.canvas_add('navaids');
+        this.canvas_add(CANVAS_NAME.STATIC);
+        this.canvas_add(CANVAS_NAME.DYNAMIC);
     }
 
     /**
+     * Add a `canvas` element to the DOM
+     *
      * @for CanvasController
-     * @method canvas_adjust_hidpi
+     * @method canvas_add
+     * @param name {CANVAS_NAME|string}
      */
-    canvas_adjust_hidpi() {
-        const dpr = window.devicePixelRatio || 1;
-
-        log(`devicePixelRatio: ${dpr}`);
-
-        // TODO: change to early return
-        if (dpr <= 1) {
-            return;
-        }
-
-        // TODO: cache this selector, $hidefCanvas
-        // TODO: replace selector with constant
-        const hidefCanvas = $(SELECTORS.DOM_SELECTORS.NAVAIDS_CANVAS).get(0);
-        const w = this.canvas.size.width;
-        const h = this.canvas.size.height;
-
-        $(hidefCanvas).attr('width', w * dpr);
-        $(hidefCanvas).attr('height', h * dpr);
-        $(hidefCanvas).css('width', w);
-        $(hidefCanvas).css('height', h);
+    canvas_add(name) {
+        const canvasTemplate = `<canvas id='${name}-canvas'></canvas>`;
 
-        const ctx = hidefCanvas.getContext('2d');
+        $(SELECTORS.DOM_SELECTORS.CANVASES).append(canvasTemplate);
```

Comment:
can we do this with a property created on instantiation from `$element` rather than using jQuery here?

Reply:
Yes we can!  Good catch. 

In the future I think we can just find the elements in the DOM instead of creating them like this. I'd like to move away from dynamically creating DOM elements unless we absolutely have to.